pandas/core/interval.py

```diff
@@ -0,0 +1,413 @@
+        # what it should actually do here...
+        loc = self.get_loc(key)  # nb. this can't handle slice objects
+        return series.iloc[loc]
+
+    def get_indexer(self, target, method=None, limit=None):
+        self._check_method(method)
+        target = _ensure_index(target)
+
+        if self.is_non_overlapping_monotonic:
+            start, stop = self._find_non_overlapping_monotonic_bounds(target)
+
+            start_plus_one = start + 1
+            if (start_plus_one < stop).any():
+                raise ValueError('indexer corresponds to non-unique elements')
+            return np.where(start_plus_one == stop, start, -1)
+
+        else:
+            if isinstance(target, IntervalIndex):
+                raise NotImplementedError(
+                    'have not yet implemented get_indexer '
+                    'for IntervalIndex indexers')
+            else:
+                return self._tree.get_indexer(target)
+
+    def delete(self, loc):
+        new_left = self.left.delete(loc)
+        new_right = self.right.delete(loc)
+        return type(self)(new_left, new_right, self.closed, self.name,
+                          fastpath=True)
+
+    def insert(self, loc, item):
+        if not isinstance(item, Interval):
+            raise ValueError('can only insert Interval objects into an '
+                             'IntervalIndex')
+        if not item.closed == self.closed:
+            raise ValueError('inserted item must be closed on the same side '
+                             'as the index')
+        new_left = self.left.insert(loc, item.left)
+        new_right = self.right.insert(loc, item.right)
+        return type(self)(new_left, new_right, self.closed, self.name,
+                          fastpath=True)
+
+    def _as_like_interval_index(self, other, error_msg):
+        self._assert_can_do_setop(other)
+        other = _ensure_index(other)
+        if (not isinstance(other, IntervalIndex) or
+                self.closed != other.closed):
+            raise ValueError(error_msg)
+        return other
+
+    def append(self, other):
+        msg = ('can only append two IntervalIndex objects that are closed on '
+               'the same side')
+        other = self._as_like_interval_index(other, msg)
+        new_left = self.left.append(other.left)
+        new_right = self.right.append(other.right)
+        if other.name is not None and other.name != self.name:
+            name = None
+        else:
+            name = self.name
+        return type(self)(new_left, new_right, self.closed, name,
+                          fastpath=True)
+
+    def take(self, indexer, axis=0):
+        indexer = com._ensure_platform_int(indexer)
+        new_left = self.left.take(indexer)
+        new_right = self.right.take(indexer)
+        return type(self)(new_left, new_right, self.closed, self.name,
+                          fastpath=True)
+
+    def __contains__(self, key):
+        try:
```

Comment:
this is the same as MultiIndex. Let's try to share some things, if you would create a mix-in (and IntervalIndex/MultiIndex can inherit)